### DISCUSSION TITLE
Handling errors and UseExceptionHandler

### DISCUSSION THREAD

**Discussion by iant-ee** [dotnet/yarp#2082] [Q&A] [answered]
I'm using Yarp to upgrade my .Net Framework MVC application to .Net 6. I'm experienced with .Net Framework, but new to .Net 6.

My application currently calls UseExceptionHandler in Program.cs, and this generally works well for us as we can examine the exception and decide what we want to log and what we want to display to the user.

However, when my .Net Framework app is not responding, it doesn't get caught by the exception handler. Instead I get a blank HTTP 502 response.

https://github.com/microsoft/reverse-proxy/issues/471 suggests using StatusCodePages but if I do that it overrides my exception handler.

How should I handle Yarp errors in an application that already calls UseExceptionHandler?

**Comment by Tratcher** [comment, marked as answer]
YARP does not throw exceptions for issues with external entities, it sets a 5xx response and adds additional info to the HttpContext. You can inspect that info to generate a custom response. See: https://microsoft.github.io/reverse-proxy/articles/middleware.html#error-handling

**Reply by Tratcher** [reply]
Also, StatusCodePages does not interfere with exception handlers, it only works with responses that did not throw or provide a body.

**Reply by iant-ee** [reply]
Thanks. Middleware didn't mean much to me, but I can see there are lots of powerful options in there. I'll take another look at why StatusCodePages is capturing the error page generated by our exception handler, but that's off-topic for here. Using the middleware documentation I've got this working for us.

For the benefit of anyone else who comes across this thread, what I've done is replace the current single line in our Program.cs

`app.MapReverseProxy();`

with 

```
app.UseEndpoints(endpoints =>  
{ 
    endpoints.MapReverseProxy(proxyPipeline => 
    { 
        // Start error handling
        proxyPipeline.Use(async (context, next) => 
        { 
            await next(); 
 
            var errorFeature = context.GetForwarderErrorFeature(); 
            if (errorFeature is not null) 
            { 
                throw errorFeature.Exception; 
            } 
        }); 
        // End error handling
        proxyPipeline.UseSessionAffinity(); 
        proxyPipeline.UseLoadBalancing(); 
        proxyPipeline.UsePassiveHealthChecks(); 
    });  
});
```

Much of that code is copied straight from the parameterless MapReverseProxy() method, which we need to do because we're modifying it (see the middleware link in @Tratcher's answer).

I've put comments around the section that deals with error handling. This simply rethrows the existing exception, which then gets caught by our existing exception handler. You could argue about the best way of rethrowing the exception, but this is good enough for our purposes, especially as we're only planning to use YARP as an interim solution.

**Reply by Tratcher** [reply]
> You could argue about the best way of rethrowing the exception, but this is good enough for our purposes, especially as we're only planning to use YARP as an interim solution.

Yeah, re-throwing like this resets the stack trace. Better to wrap the exception with a new one.

Or don't re-throw it at all, handle it directly in this middleware.
